Added paper draft (#80)
x-tabdeveloping/turftopic: paper.md

---
title: 'Turftopic: Topic Modelling with Contextual Embeddings'
tags:
  - Python
  - topic modelling
  - sentence-transformers
  - embeddings
authors:
  - name: Márton Kardos
    orcid: 0000-0001-9652-4498
    affiliation: 1
  - name: Kenneth C. Enevoldsen
    orcid: 0000-0001-8733-0966
    affiliation: 1
  - name: Jan Kostkan
    orcid: 0000-0002-9707-7121
    affiliation: 1
  - name: Ross Deans Kristensen-McLachlan
    orcid: 0000-0001-8714-1911
    affiliation: "1, 3"
  - name: Roberta Rocca
    orcid: 0000-0002-0680-7097
    affiliation: 2
affiliations:
 - name: Center for Humanities Computing, Aarhus University, Denmark
   index: 1
 - name: Interactive Minds Center, Aarhus University, Denmark
   index: 2
 - name: Department of Linguistics, Cognitive Science, and Semiotics, Aarhus University, Denmark
   index: 3
date: 17 March 2025
bibliography: paper.bib
---

# Summary

Turftopic is a topic modelling library including a number of recent topic models that go beyond bag-of-words and can understand text in context, utilizing representations from transformers.

> The library focuses on ease-of-use, providing a unified, interface for a number of different modern topic models, and boasting both model-specific and model-agnostic interpretation and visualization utilities.

Reviewer: ```suggestion
Turfttopic focuses on ease-of-use, providing a unified, interface for a number of different modern topic models, and boasting both model-specific and model-agnostic interpretation and visualization utilities.
```

The user is afforded great flexibility in model choice and customization, but the library comes with reasonable defaults, not to overwhelm first-time users with a plethora of choices.

> In addition, our library is capable of modeling topics over time, modeling topics in streams of texts, finding hierarchical themes, and multilingual usage.

Reviewer: ```suggestion
In addition, `turfttopic` allows modelling of topics over time, modeling topics in streams of texts, finding hierarchical themes, and multilingual usage.
```

consistency

> For ease of interpretation, users can utilize large language models to assign human-readable names to topics, and can model key phrases, lemmas or stems right out of the box.

Reviewer: The last part here is a bit unclear to the uninitiated reader - would reformulate

![An Overview of Turftopic's Functionality](assets/paper_banner.png)

# Statement of need

> While a number of software packages have been developed for contextual topic modelling in recent years, including BERTopic [@bertopic_paper], Top2Vec [@top2vec], CTM [@ctm], these packages include implementations of one or two topic models, and most of the utilities they provide are model-specific. This has resulted in the unfortunate situation that practitioners need to switch between different topic modelling libraries if they intend to try different models for their use case, and adapt to their particularities in both interface and functionality.

Reviewer: ```suggestion
While a number of software packages have been developed for contextual topic modelling, including BERTopic [@bertopic_paper], Top2Vec [@top2vec], CTM [@ctm], these packages include implementations of one or two topic models, and most of the utilities they provide are model-specific. This has resulted in the unfortunate situation that practitioners need to switch between different libraries and need to adapt to their particularities in both interface and functionality.
```

> Some attempts have been made at creating unified packages for modern topic models, including STREAM [@stream] and TopMost [@topmost]. These packages, however have a focus on neural models and topic model evaluation, have abstract and highly specialized interfaces, and do not include all popular topic models.

Reviewer: ```suggestion
Some attempts have been made at creating unified packages for modern topic models, including STREAM [@stream] and TopMost [@topmost]. These packages, however, have a focus on neural models and topic model evaluation, have abstract and highly specialized interfaces, and do not include popular topic models.
```

> Additionally, we considered the interpretation utilities included in these libraries fairly limited.

Reviewer: I would add here that while some of these implement advanced utilities such as dynamic topic models.
Many of these are only included in a few of the libraries locking their use behind different an unfamiliar interfaces

Reviewer: I would maybe stress that interpretation is one of the most important things about these topic modelling

> Turftopic unifies state-of-the-art contextual topic models under a superset of the scikit-learn [@scikit-learn] API, which users are likely already familiar with, and can be readily included in scikit-learn workflows and pipelines.

Reviewer: ```suggestion
Turftopic unifies state-of-the-art contextual topic models under a scikit-learn [@scikit-learn] compatible API and can be readily included in scikit-learn workflows and pipelines.
```

We focused on making Turftopic first and foremost an easy-to-use library, that does not necessitate expert knowledge or excessive amounts of code to get started with, but gives great flexibility to power users.

> The library also includes three topic models, which to our knowledge only have implementations in Turftopic, these are: KeyNMF [@keynmf], S^3^ [@s3] and GMM.

Reviewer: You could stress that these are sota topic models (which I believe is fair at least for keyNMF and s^3)

Reviewer: ```suggestion
The library also includes three topic models, which to our knowledge only have implementations in Turftopic, these are: KeyNMF [@keynmf], S^3 [@s3], and GMM.
```

# Functionality

Turftopic includes a wide array of contextual topic models from the literature, these include:

> FASTopic [@fastopic], Clustering models, such as BERTopic [@bertopic_paper] and Top2Vec [@top2vec], auto-encoding topic models, like CombinedTM [@ctm] and ZeroShotTM [@zeroshot_tm], KeyNMF [@keynmf], Semantic Signal Separation [@s3] and GMM.

Reviewer: I would introduce the shorthand for s^3 and then re-use it throughout.

> We believe these models to be representative of the state of the art in contextual topic modelling and intend to expand on them in the future.

Reviewer: ```suggestion
These models represent the current state of the art in contextual topic modelling, but we intend to expand the package in response to new releases.
```

![Components of a Topic Modelling Pipeline in Turftopic](https://x-tabdeveloping.github.io/turftopic/images/topic_modeling_pipeline.png){width="800px"}

Each model in Turftopic has an *encoder* component, which is used for producing continuous document-representations, and a *vectorizer* component, which extracts term counts in each documents, thereby dictating which terms will be considered in topics.
The user has full control over what components should be used at different stages of the topic modelling process, thereby having fine-grained influence on the nature and quality of topics.

> The library comes loaded with a lot of utilities to help users interpret their results, including *pretty printing* utilities for exploring topics, *interactive visualizations* partially powered by the `topicwizard` [@topicwizard] Python package, and *automated topic naming* with LLMs.

Reviewer: You never introduced LLMs

To accommodate a variety of use-cases, Turftopic can be used for dynamic topic modelling, where we expect topics to change over time, can be used for uncovering hierarchical structure in topics.
Some models can also be fitted in an *online* fashion, where documents are accounted for as they come in by batches.
Turftopic also includes *seeded* topic modelling, where a seed phrase can be used to retrieve topics relevant to the specific research question.

# Target Audience

Topic models can be utilized in a number of research settings, including exploratory data analysis, discourse analysis of many domains, such as newspapers, social media or policy documents.
Turftopic has already been utilized by @keynmf for analyzing information dynamics in Chinese Diaspora Media, and is currently being used in multiple ongoing research projects, including one analyzing discourse on the HPV vaccine in Denmark.

> In addition, the library has already seen extensive use in student exam papers at Aarhus University.

Reviewer: ```suggestion
```

> We expect that computational researchers, especially in digital humanities and in social sciences will benefit from an easy-to-use, unified contextual topic modelling package.

Reviewer: I would move this to the beginning (target audience -> more specific use setting -> exact usage with citations). 

I would also make sure to include industry here

(this section is as much as who we want to target not who are currently using it)

---
title: 'Turftopic: Topic Modelling with Contextual Representations from Sentence Transformers'
tags:
  - Python
  - topic modelling
  - sentence-transformers
  - embeddings
authors:
  - name: Márton Kardos
    orcid: 0000-0001-9652-4498
    affiliation: 1
  - name: Kenneth C. Enevoldsen
    orcid: 0000-0001-8733-0966
    affiliation: 1
  - name: Jan Kostkan
    orcid: 0000-0002-9707-7121
    affiliation: 1
  - name: Ross Deans Kristensen-McLachlan
    orcid: 0000-0001-8714-1911
    affiliation: "1, 3"
  - name: Roberta Rocca
    orcid: 0000-0002-0680-7097
    affiliation: 2
affiliations:
 - name: Center for Humanities Computing, Aarhus University, Denmark
   index: 1

>  - name: Interactive Minds Center, Aarhus University, Denmark

Reviewer 2: Interacting

   index: 2
 - name: Department of Linguistics, Cognitive Science, and Semiotics, Aarhus University, Denmark
   index: 3
date: 17 March 2025
bibliography: paper.bib
---

# Summary

Turftopic is a topic modelling library including a number of recent topic models that go beyond bag-of-words and can understand text in context, utilizing representations from transformers.

> Turftopic focuses on ease-of-use, providing a unified, interface for a number of different modern topic models, and boasting both model-specific and model-agnostic interpretation and visualization utilities.

Reviewer 2: should be "ease of use"

The user is afforded great flexibility in model choice and customization, but the library comes with reasonable defaults, not to overwhelm first-time users with a plethora of choices.

> In addition, Turftopic allows you to model topics, as they change over time, learning themes from streams of texts, finding hierarchical topics, and multilingual usage.

Reviewer 2: the syntax here is a bit funky, maybe: "Turftopic allows you to model topics as they change over time by learning themes from streams of texts, to find hierarchical topics, and to learn topics in multilingual texts and corpora."

Reviewer 3: I'd actually even suggest a numbered list for sake of readability. But this is just a stylistic preference I have, Roberta's suggestion also works.

_Turftopic allows you to: 1) model topics as they change over time by learning themes from streams of texts; 2) find hierarchical topics; and 3) learn topics in multilingual texts and corpora._

Users can utilize the power of large language models (LLMs) to give human-readable names to topics.
Turftopic also comes with built-in utilities for generating topic descriptions based on key-phrases or lemmas rather than individual words.

![An Overview of Turftopic's Functionality](assets/paper_banner.png)

# Statement of Need

While a number of software packages have been developed for contextual topic modelling in recent years, including BERTopic [@bertopic_paper], Top2Vec [@top2vec], CTM [@ctm], these packages include implementations of one or two topic models, and most of the utilities they provide are model-specific. This has resulted in the unfortunate situation that practitioners need to switch between different libraries and adapt to their particularities in both interface and functionality.
Some attempts have been made at creating unified packages for modern topic models, including STREAM [@stream] and TopMost [@topmost].

> These packages, however have a focus on neural models and topic model evaluation, have abstract and highly specialized interfaces, and do not include some popular topic models.

Reviewer 2: v minor, but comma after however, or shift it to start sentence

Additionally, while model interpretation is an incredibly important aspect of topic modelling, the interpretation utilities provided in these libraries are fairly limited, especially in comparison with model-specific packages, like BERTopic.

Turftopic unifies state-of-the-art contextual topic models under a superset of the `scikit-learn` [@scikit-learn] API, which users are likely already familiar with, and can be readily included in `scikit-learn` workflows and pipelines.
We focused on making Turftopic first and foremost an easy-to-use library, that does not necessitate expert knowledge or excessive amounts of code to get started with, but gives great flexibility to power users.

> Furthermore, included an extensive suite of pretty-printing and visualization utilities that aid users in interpreting their results.

Reviewer 2: *we included

> The library also includes three topic models, which to our knowledge only have implementations in Turftopic, these are: KeyNMF [@keynmf], S^3^ [@s3], and GMM.

Reviewer 2: I'd spell out GMM and give a very quick explanation if there is no reference for it

Reviewer 3: Agree with this

# Functionality

Turftopic includes a wide array of contextual topic models from the literature, these include:
FASTopic [@fastopic], Clustering models, such as BERTopic [@bertopic_paper] and Top2Vec [@top2vec], auto-encoding topic models, like CombinedTM [@ctm] and ZeroShotTM [@zeroshot_tm], KeyNMF [@keynmf], Semantic Signal Separation [@s3] and GMM.

> We believe these models to be representative of the state of the art in contextual topic modelling and intend to expand on them in the future.

Reviewer 2: I'd probably remove the "we believe"

![Components of a Topic Modelling Pipeline in Turftopic](https://x-tabdeveloping.github.io/turftopic/images/topic_modeling_pipeline.png){width="800px"}

> Each model in Turftopic has an *encoder* component, which is used for producing continuous document-representations, and a *vectorizer* component, which extracts term counts in each documents, thereby dictating which terms will be considered in topics.

Reviewer 3: Do we need to explain what an encoder is? Or what transformers are? I think citing Vaswani, et al. (2017) is probably enough tbh.

The user has full control over what components should be used at different stages of the topic modelling process, thereby having fine-grained influence on the nature and quality of topics.

The library comes loaded with a lot of utilities to help users interpret their results, including *pretty printing* utilities for exploring topics, *interactive visualizations* partially powered by the `topicwizard` [@topicwizard] Python package, and *automated topic naming* with LLMs.

> To accommodate a variety of use-cases, Turftopic can be used for dynamic topic modelling, where we expect topics to change over time, can be used for uncovering hierarchical structure in topics.

Reviewer 2: missing an "and it" after "change over time" I think.  also no dash in use-cases. What a "hierarchical" topic is is never explained I think - I know it's semi-intuitive, but you could simply explain it is about simultaneous extraction of topics and multiple levels of granularity?

> Some models can also be fitted in an *online* fashion, where documents are accounted for as they come in by batches.

Reviewer 2: "_in_ batches"?

Turftopic also includes *seeded* topic modelling, where a seed phrase can be used to retrieve topics relevant to the specific research question.

# Use Cases

> Topic models can be utilized in a number of research settings, including exploratory data analysis, discourse analysis of diverse domains, such as newspapers, social media or policy documents.

Reviewer 2: I'd go bolder here and say something along the lines of "topic modeling is a key tool for quantitative text analysis" and you could cite nielbo (https://pure.uva.nl/ws/files/169498673/s43586-024-00302-w.pdf) who includes it in table 1, and/or borrow from the very first lines of the S3 paper.

Reviewer 3: +1

> Turftopic has already been utilized by @keynmf for analyzing information dynamics in Chinese Diaspora Media, and is currently being used in multiple ongoing research projects, including one analyzing discourse on the HPV vaccine in Denmark, and studying Danish golden-age literature.

Reviewer 2: Just to have one more example that has a citation (and something that showcases its utility for ML specialists), you could also reference S3 and say that its ease of use and unified interface make it possible to compare topic modeling approaches at scale (I know it's topic-benchmark that makes it really easy, but still)

# Target Audience

We expect that Turftopic will prove useful to a diverse user base including computational researchers in digital humanities and social sciences, and industry NLP professionals.

> Due to ease of use, Turftopic is also an appropriate choice for educational purposes.

Reviewer 3: "[...] for educational purposes, providing instructors with a single, user-friendly framework for students to explore and compare alternative topic modelling approaches".

(use "modeling" if we're speaking American)

To accommodate a variety of use cases, Turftopic can be used for *dynamic* topic modelling, where we expect topics to change over time.
Turftopic is also capable of extracting topics at multiple levels of granularity, thereby uncovering *hierarchical* topic structures.
Some models can also be fitted in an *online* fashion, where documents are accounted for as they come in batches.
Turftopic also includes *seeded* topic modelling, where a seed phrase can be used to retrieve topics relevant to the specific research question.

# Use Cases

Topic modelling is a key tool for quantitative text analysis [@quantitative_text_analysis], and can be utilized in a number of research settings, including exploratory data analysis, discourse analysis of diverse domains, such as newspapers, social media or policy documents.
Turftopic has already been utilized by @keynmf for analyzing information dynamics in Chinese diaspora media, and is currently being used in multiple ongoing research projects, including one analyzing discourse on the HPV vaccine in Denmark, and studying Danish golden-age literature.

# Target Audience

> We expect that Turftopic will prove useful to a diverse user base including computational researchers in digital humanities and social sciences, and industry NLP professionals.

Reviewer: ```suggestion
We expect that Turftopic will prove useful to a diverse user base across both industry and research including researchers in digital humanities and social sciences, and NLP professionals.
```

> Turftopic is also an appropriate choice for educational purposes, providing instructors with a single, user-friendly framework for students to explore and compare alternative topic modelling approaches.

Reviewer: instructions?
